cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/CombinedMetricsScanner.java

```diff
@@ -0,0 +1,178 @@
+import co.cask.cdap.api.dataset.table.Row;
+import co.cask.cdap.api.dataset.table.Scanner;
+import com.google.common.annotations.VisibleForTesting;
+
+import java.util.Iterator;
+import java.util.Map;
+import java.util.NoSuchElementException;
+import java.util.TreeMap;
+import javax.annotation.Nullable;
+
+/**
+ * Implements Scanner on merge data from v2 and v3 metrics table.
+ */
+public class CombinedMetricsScanner implements Scanner {
+
+  private final Scanner v2TableScanner;
+  private final Scanner v3TableScanner;
+  private Row v2TableNextRow;
+  private Row v3TableNextRow;
+
+  @VisibleForTesting
+  public CombinedMetricsScanner(Scanner v2TableScanner, Scanner v3TableScanner) {
+    this.v3TableScanner = v3TableScanner;
+    this.v2TableScanner = v2TableScanner;
+    this.v3TableNextRow = v3TableScanner.next();
+    this.v2TableNextRow = v2TableScanner.next();
+  }
+
+  @Nullable
+  @Override
+  public Row next() {
+    Row resultRow;
+    // both the scanners are exhausted, so return null
+    if (v2TableNextRow == null && v3TableNextRow == null) {
+      return null;
+    }
+
+    // v2 scanner is exhausted, so return row from v3 table
+    if (v2TableNextRow == null) {
+      return getV3Row();
+    }
+
+    // v3 scanner is exhausted, so return row from v2 table
+    if (v3TableNextRow == null) {
+      return getV2Row();
+    }
+
+    // if both the scanners have rows, return the minimum row.
+    int rowComparison = Bytes.compareTo(v2TableNextRow.getRow(), v3TableNextRow.getRow());
+
+    if (rowComparison > 0) {
+      resultRow = getV3Row();
+    } else if (rowComparison < 0) {
+      resultRow = getV2Row();
+    } else {
+      // if the rows are same from v2 and v3 tables, we will merge the columns from both the tables and return the
+      // results. For example,
+
+      // v2 table has row1: t1 t2 t3 t4 t5   and v3 table has row1: t4 t5 t6 t7 t8
+      //                    v1 v2 v3 v4 v5                          v6 v7 v8 v9 v10
+
+      // then we will return row1: t1   t2   t3   t4     t5    t6   t7   t8
+      //                           v1   v2   v3  v4+v6  v5+v7  v8   v9   v10
+
+      Map<byte[], byte[]> mergedColumns = new TreeMap<>(Bytes.BYTES_COMPARATOR);
+      byte[] rowKey = v3TableNextRow.getRow();
+
+      mergeColumns(v2TableNextRow, v3TableNextRow, mergedColumns);
+
+      v2TableNextRow = v2TableScanner.next();
+      v3TableNextRow = v3TableScanner.next();
+      resultRow = new Result(rowKey, mergedColumns);
+    }
+
+    return resultRow;
+  }
+
+  @Override
+  public void close() {
+    v3TableScanner.close();
+    v2TableScanner.close();
+  }
+
+  private void mergeColumns(Row v2TableNextRow, Row v3TableNextRow, Map<byte[], byte[]> mergedColumns) {
+    Iterator<Map.Entry<byte[], byte[]>> v2ColumnIterator = v2TableNextRow.getColumns().entrySet().iterator();
+    Iterator<Map.Entry<byte[], byte[]>> v3ColumnIterator = v3TableNextRow.getColumns().entrySet().iterator();
+
```

Comment:
seems that it would be a lot simpler to 
1. copy v2 row into a new Map
2. iterate over v3 row and for each column:
- if it already exists, convert both values to long, add them and write the sum back as bytes
- otherwise copy this column to the new map.